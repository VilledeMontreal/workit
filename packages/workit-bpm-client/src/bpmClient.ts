/*
 * Copyright (c) 2020 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { Client as CamundaExternalClient } from 'camunda-external-task-client-js';

export { CamundaExternalClient };
