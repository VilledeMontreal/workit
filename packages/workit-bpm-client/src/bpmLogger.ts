/*
 * Copyright (c) 2020 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { logger as camundaLogger } from 'camunda-external-task-client-js';

export { camundaLogger };
