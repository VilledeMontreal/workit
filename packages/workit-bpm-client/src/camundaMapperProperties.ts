/*
 * Copyright (c) 2020 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { IVariablePayload, IWorkflowProps } from 'workit-types';

export class CamundaMapperProperties {
  public static map(task: IVariablePayload): IWorkflowProps {
    return {
      activityId: task.activityId,
      businessKey: task.businessKey,
      processInstanceId: task.processInstanceId,
      workflowDefinitionVersion: Number(task.processDefinitionId.split(':')[1]),
      workflowInstanceKey: task.processDefinitionId,
      workflowKey: task.processDefinitionKey,
      bpmnProcessId: task.processDefinitionKey,
      customHeaders: CamundaMapperProperties._getCustomHeaders(task),
      jobKey: task.id,
      retries: task.retries,
      topicName: task.topicName,
      workerId: task.workerId,
      lockExpirationTime: new Date(task.lockExpirationTime)
    };
  }

  public static unmap(props: IWorkflowProps): IVariablePayload {
    throw new Error('Not Implemented yet');
  }

  private static _getMeta(task: IVariablePayload) {
    return task.variables.get('_meta');
  }

  private static _getCustomHeaders(task: IVariablePayload) {
    const meta = CamundaMapperProperties._getMeta(task);
    if (meta && meta.customHeaders) {
      return meta.customHeaders;
    }
    return {};
  }
}
