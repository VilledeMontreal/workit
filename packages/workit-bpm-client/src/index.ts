/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

// IOC
export * from './config/constants/identifiers';
import './config/container';

export * from './camundaBpmClient';
export * from './bpmClient';
export * from './bpmLogger';
export * from './camundaBpmClient';
export * from './camundaMapperProperties';
export * from './logger';
export * from './camundaMessage';

export * from './utils/utils';
export * from './variables';
