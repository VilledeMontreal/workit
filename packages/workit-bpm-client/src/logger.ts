/*
 * Copyright (c) 2020 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

/* eslint @typescript-eslint/no-unsafe-assignment: 0 */ // --> OFF
/* eslint @typescript-eslint/ban-types: 0 */ // --> OFF
/* eslint @typescript-eslint/restrict-template-expressions: 0 */ // --> OFF
/* eslint @typescript-eslint/no-unsafe-call: 0 */ // --> OFF
/* eslint @typescript-eslint/no-unsafe-member-access: 0 */ // --> OFF
// eslint-disable-next-line @typescript-eslint/no-var-requires
const stringify = require('fast-safe-stringify');

const { pid } = process;
const { platform } = process;

/**
 * @returns a formatted success message
 */
const success = (message: string | {}) =>
  process.stdout.write(
    `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":30,"msg":"${stringify(
      message
    )}","time":"${new Date().toISOString()}","v":0}\n`
  );

/**
 * @returns a formatted error message
 */
const error = (message: string | {}) =>
  process.stdout.write(
    `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":50,"msg":"${stringify(
      message
    )}","time":"${new Date().toISOString()}","v":0}\n`
  );

/**
 * logs various events from client
 * @param client
 */
const log = (client: any) => {
  client.on('subscribe', (topic: string) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","platform":"${platform}","logType":"mtl-worker","level":30,"msg":"subscribed to topic ${topic}","time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('unsubscribe', (topic: string) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","platform":"${platform}","logType":"mtl-worker","level":40,"msg":"unsubscribed from topic ${topic}","time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  // client.on('poll:start', () => {
  //   console.log(
  //     `{"name":"default","pid":"${pid}","platform":"${platform}","logType":"mtl-worker","level":40,"msg":"start polling","time":"${new Date().toISOString()}","v":0}`
  //   );
  // });

  client.on('poll:stop', () => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","platform":"${platform}","logType":"mtl-worker","level":40,"msg":"stop polling","time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  //   client.on("poll:success", tasks => {
  //     const output = success(`polled ${tasks.length} tasks`);
  //     console.log(output);
  //   });

  client.on('poll:error', (e: Error) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":50,"msg":"polling failed","error":${stringify(
        e
      )},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('complete:success', (task: any) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":30,"msg":"task completed","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('complete:error', (task: any, e: Error) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":50,"msg":"task failed","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"retries":"${task.retries}","error":"${stringify(e)}","time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('handleFailure:success', (task: any) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":40,"msg":"handle failure completed","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('handleFailure:error', (task: any, e: Error) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":50,"msg":"handle failure failed... Happy coding!","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"retries":"${task.retries}","error":${stringify(e)},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('handleBpmnError:success', (task: any) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":40,"msg":"handle Bpmn error completed","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('handleBpmnError:error', (task: any, e: Error) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":50,"msg":"handle bpmn error failed... Happy coding!","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"retries":"${task.retries}","error":${stringify(e)},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('extendLock:success', (task: any) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":30,"msg":"extend lock success","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('extendLock:error', (task: any, e: Error) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":50,"msg":"extend lock error","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"retries":"${task.retries}","error":${stringify(e)},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('unlock:success', (task: any) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":30,"msg":"unlock success","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });

  client.on('unlock:error', (task: any, e: Error) => {
    process.stdout.write(
      `{"name":"default","pid":"${pid}","logType":"mtl-worker","level":50,"msg":"unlock error","activityId":"${
        task.activityId
      }","processInstanceId":"${task.processInstanceId}","workerId":"${task.workerId}","topicName":"${
        task.topicName
      }","processDefinitionId":"${task.processDefinitionId}","variables":${stringify(
        task.variables.getAll()
      )},"retries":"${task.retries}","error":${stringify(e)},"time":"${new Date().toISOString()}","v":0}\n`
    );
  });
};

export const logger = Object.assign(log, { success, error });
