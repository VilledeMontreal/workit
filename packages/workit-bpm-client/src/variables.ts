/*
 * Copyright (c) 2020 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { Variables } from 'camunda-external-task-client-js';

export { Variables };
