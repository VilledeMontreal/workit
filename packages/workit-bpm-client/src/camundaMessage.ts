/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { getVariablesWhenChanged, ProxyFactory } from 'workit-core';
import {
  FailureException,
  ICamundaService,
  IMessage,
  IVariablePayload,
  IVariables,
  IWorkflowProps
} from 'workit-types';
import { CamundaMapperProperties } from './camundaMapperProperties';
import { Variables } from './variables';

const stringify = require('fast-safe-stringify');

export class CamundaMessage {
  public static wrap(payload: { task: IVariablePayload; taskService: any }): [IMessage, ICamundaService] {
    const { task } = payload;
    const properties = CamundaMapperProperties.map(task);
    const messageWithoutSpan = {
      body: task.variables.getAll(),
      properties: properties as IWorkflowProps
    };
    // TODO: create a CamundaMessage builder
    const msg = ProxyFactory.create({
      body: messageWithoutSpan.body,
      properties: messageWithoutSpan.properties
    });
    return [
      msg,
      // TODO: create a CamundaService builder
      {
        hasBeenThreated: false,
        /**
         * Acknowledge the message to Camunda platform
         * Variables will be updated if change has been detected
         */
        async ack(message: IMessage) {
          if (this.hasBeenThreated) {
            return;
          }

          const vars = getVariablesWhenChanged<IVariables>(message, m => CamundaMessage.unwrap(m));

          await payload.taskService.complete(task, vars);
          this.hasBeenThreated = true;
        },
        /**
         * Un acknowledge the message to Camunda platform
         * This will handle failure.
         * Notice that on failure, the current camunda client doesn't update the variables
         */
        async nack(error: FailureException) {
          if (this.hasBeenThreated) {
            return;
          }
          const retries = error.retries;
          await payload.taskService.handleFailure(task, {
            errorMessage: error.message,
            errorDetails: stringify(error),
            retries,
            // TODO: Add to configuration
            retryTimeout: 1000 * retries * 2
          });
          this.hasBeenThreated = true;
        }
      }
    ];
  }

  public static unwrap(message: IMessage): IVariables {
    const body = message.body;
    const vars = new Variables(body);
    for (const key in body) {
      if (Object.prototype.toString.call(body[key]) === '[object Date]') {
        // Otherwise, we got invalid date
        // TODO: Check the cause
        vars.setTyped(key, { type: 'date', value: (body[key] as Date).toUTCString(), valueInfo: {} });
      } else {
        vars.set(key, body[key]);
      }
    }
    CamundaMessage._setCustomHeaders(vars, message.properties.customHeaders);
    return vars;
  }

  private static _setCustomHeaders(vars: IVariables, customHeaders: any) {
    if (customHeaders && Object.keys(customHeaders).length > 0) {
      vars.set('_meta', { customHeaders });
    }
  }
}
