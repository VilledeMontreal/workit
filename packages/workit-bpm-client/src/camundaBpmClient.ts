/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { PaginationUtils } from './utils/paginationUtils';

import {
  IBpmn,
  ICamundaClient,
  ICamundaConfig,
  ICamundaRepository,
  ICamundaService,
  IClient,
  ICreateWorkflowInstance,
  ICreateWorkflowInstanceResponse,
  IDeployWorkflowResponse,
  IHttpResponse,
  IMessage,
  IPagination,
  IPaginationOptions,
  IPublishMessage,
  ITopicSubscription,
  IUpdateWorkflowRetry,
  IUpdateWorkflowVariables,
  IVariablePayload,
  IWorkflow,
  IWorkflowClient,
  IWorkflowDefinition,
  IWorkflowDefinitionRequest,
  IWorkflowOptions,
  IWorkflowProcessIdDefinition
} from 'workit-types';
import { CamundaMessage } from './camundaMessage';
import { CamundaRepository } from './repositories/camundaRepository';
export class CamundaBpmClient implements IClient<ICamundaService>, IWorkflowClient {
  private static _getWorkflowParams(options?: Partial<IWorkflowOptions & IPaginationOptions>): any {
    const _params = {} as any;
    if (options && (options as IWorkflowOptions).bpmnProcessId) {
      _params.key = options.bpmnProcessId;
    }
    return PaginationUtils.setCamundaBpmPaginationParams(_params, options);
  }
  private readonly _client: ICamundaClient;
  private _topicSubscription: ITopicSubscription | undefined;
  private readonly _config: ICamundaConfig;
  private readonly _repo: ICamundaRepository;
  constructor(config: ICamundaConfig, client: ICamundaClient) {
    this._client = client;
    this._config = config;
    this._repo = new CamundaRepository(config);
  }
  public subscribe(onMessageReceived: (message: IMessage, service: ICamundaService) => Promise<void>): Promise<void> {
    this._topicSubscription = this._client.subscribe(
      this._config.topicName,
      this._config.subscriptionOptions,
      async (camundaObject: { task: IVariablePayload; taskService: any }) => {
        const [message, service] = CamundaMessage.wrap(camundaObject);
        await onMessageReceived(message, service);
      }
    );

    this._startSubscriber();

    return Promise.resolve();
  }
  public unsubscribe(): Promise<void> {
    try {
      if (this._topicSubscription) {
        this._topicSubscription.unsubscribe();
      }
      this._client.stop();
      return Promise.resolve();
    } catch (error) {
      return Promise.reject(error);
    }
  }
  public async deployWorkflow(absPath: string): Promise<IDeployWorkflowResponse> {
    const result = await this._repo.deployWorkflow(`Deploy from ${this._config.workerId}`, absPath);
    const response = result.data;
    const deployedProcessDefinitionsId = Object.keys(response.deployedProcessDefinitions)[0];
    const definition = response.deployedProcessDefinitions[deployedProcessDefinitionsId];
    const workflows = [
      {
        bpmnProcessId: definition.key,
        workflowKey: definition.id,
        resourceName: definition.resource,
        version: definition.version
      }
    ];
    return {
      workflows,
      key: response.id
    };
  }
  public async getWorkflows(options?: Partial<IWorkflowOptions & IPaginationOptions>): Promise<IPagination<IWorkflow>> {
    const params = CamundaBpmClient._getWorkflowParams(options);
    const apiOptions = { params };
    const requests: [
      Promise<IHttpResponse<IBpmn[]>>,
      Promise<
        IHttpResponse<{
          count: number;
        }>
      >
    ] = [this._repo.getWorkflows(apiOptions), this._repo.getWorkflowCount(apiOptions)];
    const [result, repCount] = await Promise.all(requests);
    const bpmns = result.data as IBpmn[];
    const workflows = bpmns.map(definition => {
      return {
        bpmnProcessId: definition.key,
        workflowKey: definition.id,
        resourceName: definition.resource,
        version: definition.version
      };
    });

    return {
      paging: PaginationUtils.getPagingFromOptions(repCount.data.count, options),
      items: workflows
    };
  }
  public async getWorkflow(payload: IWorkflowDefinitionRequest): Promise<IWorkflowDefinition> {
    let definition;
    if (this._hasBpmnProcessId(payload)) {
      definition = await this._repo.getWorkflow(payload.bpmnProcessId);
    } else {
      definition = await this._repo.getWorkflow(payload.workflowKey);
    }

    return {
      bpmnProcessId: definition.key,
      bpmnXml: definition.bpmn20Xml,
      resourceName: definition.resource,
      version: definition.version,
      workflowKey: definition.id
    };
  }
  public async updateVariables<T = any>(model: IUpdateWorkflowVariables<Partial<T>>): Promise<void> {
    await this._repo.updateVariables(model.processInstanceId, model.variables);
  }
  public async updateJobRetries({ jobKey, retries }: IUpdateWorkflowRetry): Promise<void> {
    await this._repo.updateJobRetries(jobKey, retries);
  }
  public publishMessage<T, K>(payload: IPublishMessage<T, K>): Promise<void> {
    return this._repo.publishMessage({
      messageName: payload.name,
      processInstanceId: payload.messageId as string,
      correlationKeys: payload.correlation,
      variables: payload.variables
    });
  }
  public async createWorkflowInstance<T>(model: ICreateWorkflowInstance<T>): Promise<ICreateWorkflowInstanceResponse> {
    const result = await this._repo.createWorkflowInstance(model.bpmnProcessId, model.variables);
    const response = result.data;
    const bpmnDef = response.definitionId.split(':');
    // TODO: fix this type issue
    return {
      bpmnProcessId: bpmnDef[0],
      version: (bpmnDef[1] as unknown) as number,
      workflowInstanceKey: response.id,
      workflowKey: response.definitionId
    };
  }
  public cancelWorkflowInstance(instanceId: string): Promise<void> {
    return this._repo.cancelWorkflowInstance(instanceId);
  }
  public resolveIncident(incidentKey: string): Promise<void> {
    return this._repo.resolveIncident(incidentKey);
  }
  private _startSubscriber() {
    if (!this._config.autoPoll) {
      this._client.start();
    }
  }
  private _hasBpmnProcessId(request: IWorkflowDefinitionRequest): request is IWorkflowProcessIdDefinition {
    return (request as IWorkflowProcessIdDefinition).bpmnProcessId !== undefined;
  }
}
