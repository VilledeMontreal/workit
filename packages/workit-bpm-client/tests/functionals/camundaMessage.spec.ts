/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { CamundaMessage } from '../../src/camundaMessage';
import { Variables } from '../../src/variables';

describe('camundaMessage', () => {
  it('unmap', () => {
    const message = {
      body: { hello: 'world' },
      properties: {
        workflowInstanceKey: '38963',
        bpmnProcessId: 'test-process',
        workflowDefinitionVersion: 4,
        workflowKey: '8806',
        activityId: 'ServiceTask_0xdwuw7',
        elementInstanceKey: '38967',
        customHeaders: { jwt: 'jwt fake' },
        jobKey: '38968',
        processInstanceId: '38963',
        retries: 1,
        lockExpirationTime: new Date(1562269537659),
        topicName: 'topic_demo',
        workerId: 'demo',
      },
    };
    const camundaObject = CamundaMessage.unwrap(message);
    expect(camundaObject).toMatchSnapshot();
  });

  it('wrap', () => {
    const camundaPayload = {
      task: { processInstanceId: '38963', processDefinitionId: 'xxxxx', variables: new Variables() } as any,
      taskService: {
        handleFailure: jest.fn(),
        complete: jest.fn(),
      },
    };
    const [, service] = CamundaMessage.wrap(camundaPayload);
    service.nack({ name: 'error', message: 'Oopps', retries: 0, retryTimeout: 15_000 });
    expect(camundaPayload.taskService.handleFailure).toBeCalledTimes(1);
    expect(camundaPayload.taskService.complete).toBeCalledTimes(0);
    expect(camundaPayload.taskService.handleFailure.mock.calls[0][1]).toMatchSnapshot();
  });
});
