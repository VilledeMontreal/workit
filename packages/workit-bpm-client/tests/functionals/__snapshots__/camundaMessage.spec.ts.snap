// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`camundaMessage unmap 1`] = `
Variables {
  "get": [Function],
  "getAll": [Function],
  "getAllTyped": [Function],
  "getDirtyVariables": [Function],
  "getTyped": [Function],
  "set": [Function],
  "setAll": [Function],
  "setAllTyped": [Function],
  "setTyped": [Function],
}
`;

exports[`camundaMessage wrap 1`] = `
Object {
  "errorDetails": "{\\"name\\":\\"error\\",\\"message\\":\\"Oopps\\",\\"retries\\":0,\\"retryTimeout\\":15000}",
  "errorMessage": "Oopps",
  "retries": 0,
  "retryTimeout": 15000,
}
`;
